### DISCUSSION TITLE
Error booting Linux EFI kernel.....

### DISCUSSION THREAD

**Discussion by cacheman** [ipxe/ipxe#902] [Q&A] [answered]
Hi....

I have run into a strange problem booting a static (i.e. no initrd) Linux 5.x or 6.x EFI kernel using iPXE 1.21.1. My menu entry is below:

`:TestLinux`
`kernel http://aaa.bbb.ccc.ddd/bzImage root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317`
`boot || goto failed`
`goto start`

The kernel is one which I have built from source. Kernel versions v5.x and v6.x both work fine with iPXE v1.20.1.....and they have worked just fine for a long time....

However, with latest v1.21.1 the **_first_** time I select this in my PXE menu I get the following error:

`http://aaa.bbb.ccc.ddd/bzImage... ok`
`EFI stub: ERROR: Failed to load initrd: 0x8000000000000001`
`EFI stub: ERROR: efi_main() failed!`
`Could not boot: Error 0x7f048281 (https://ipxe.org/7f048281)`
`Booting failed.`
`Press key to continue`

....then when I 'Press key to continue' I get presented with my PXE menu again. The really strange thing is if I now for a **_second_** time select the same menu entry it gives a different error message, but it **_DOES_** work this time:

`http://aaa.bbb.ccc.ddd/bzImage... ok`
`EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path`
`_......kernel boots with usual Linux kernel boot messages....._`

Is there something new in v1.21.1 that I'm missing or perhaps doing incorrectly?
Or possibly a kernel build option I need to enable/disable?
Or might this be a bug?

Would be grateful for any help with this.

Thanks in advance.

Cacheman...

**Comment by NiKiZe** [comment]
Are you using git master? (Which is the latest version)

At complains on initrd, but you have not loaded any initrd at all.
However I think we recently got the media_guid added, so maybe something is going on with that.

Before your boot line, could you try adding an imgstat line and post output for both runs? (I expect it to first fail and then succeed, but with something changing in the imgstat output)

**Reply by cacheman** [reply]
Hi....

Thanks for the swift reply. As regards which git branch I'm using, I think it's the master branch.....to be honest all I did was what it says to do on ipxe.org which is 'git clone https://github.com/ipxe/ipxe.git'.

Anyway, I've rerun my menu entry with imgstat....first run shows the following:

`http://aaa.bbb.ccc.ddd/bzImage... ok`
`bzImage: 9531264 bytes [EFI] [SELECTED] "root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317"`
`EFI stub: ERROR: Failed to load initrd: 0x8000000000000001`
`EFI stub: ERROR: efi_main() failed!`
`Could not boot: Error 0x7f048281 (https://ipxe.org/7f048281)`
`Booting failed.`
`Press key to continue`

....and the second run seems to just display the output of imgstat twice before booting normally....:

`http://aaa.bbb.ccc.ddd/bzImage... ok`
`bzImage: 9531264 bytes [EFI] [SELECTED] "root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317"`
`bzImage: 9531264 bytes [EFI] [SELECTED] "root=/dev/nfs nfsroot=aaa.bbb.ccc.ddd:/export/PC,vers=3,rsize=8192,wsize=8192,actimeo=9999 ip=dhcp nofb nomodeset nohugeiomap vga=0x317"`

Hope this helps.

Cacheman.

**Reply by NiKiZe** [reply]
I think the bzImage is just loaded twice, and that works.

You say there is an error message when it works? I don't see any errors.

For now I would just add a minimal initrd line with a small file to see if that solves your issue (as a temporary workaround)

**Reply by stappersg** [reply]
> You say there is an error message when it works? I don't see any errors.

From the original posting:
```text
Could not boot: Error 0x7f048281 (https://ipxe.org/7f048281)
```

And https://ipxe.org/7f048281 says **"Error: Could not start image"** and recomments _Try building iPXE with the debug option `DEBUG=efi_image`_

**Comment by mcb30** [comment, marked as answer]
I suspect this may turn out to be a quirk of what happens when booting **without** an initrd when the bootloader exposes a zero-length file via the `LINUX_EFI_INITRD_MEDIA_GUID` device path, which iPXE has supported very recently in commit https://github.com/ipxe/ipxe/commit/6a004be0c

**Reply by NiKiZe** [reply]
The zero length file will be initrd.magic correct?
Do we need to prevent it when zero size? Or should it have a valid cpio header but no files? (Which could cause different issues)

**Reply by mcb30** [reply]
Conversation moved to issue #903

**Reply by mcb30** [reply]
@cacheman Could you please let us know (via issue #903) your name and email address for the commit log?
